daemon/monster.8
.TH monster "8" "October 2004" "umweb" "System Manager's Manual"
.SH NAME
.B monster
\- cleans up cookies and tickets, propagates timestamps
and state
.SH SYNOPSIS
.B monster
[
.B \-dV
] [
.BI \-F\  syslog-facility
] [
.BI \-h\  replication-hostname
] [
.BI \-H\  hard-timeout
] [
.BI \-i\  idle-cache-timeout
] [
.BI \-I\  timestamp-pushing-interval
] [
.BI \-l\  loggedout-cache
] [
.BI \-L\  syslog-level
] [
.BI \-p\  port
] [
.BI \-x\  ca-dir
] [
.BI \-y\  cert-pem-file
] [ 
.BI \-z\  private-key-file
]
.sp
.SH DESCRIPTION
Monster serves two separate functions. In all cases, it is the mechanism by
which old cookies and tickets can be deleted when they are no longer needed.
In addition, in a replicated cosign environment, monster can be used to
propagate the time stamps and state of the login cookies.
Upon startup, monster disassociates from the terminal. This can be
prevented by using the -d option. If the -h option is give, monster
retrieves the host information for
.I replication-hostname
and opens an SSL connection to the cosignds on all of the other cosign
hosts. 
.sp
Monster sleeps for 120 seconds between successive passes through the cookie 
database ( _COSIGN_DIR ).  This interval can be changed to
an arbitrary
.I timestamp-pushing-interval
with the -I option.
Monster stats each login cookie, checking to see if it is either past the
.I idle-timeout ( 4.5 hours - idle time + grey window from cosignd, + the usual loggedout_cache time in monster)
, the 
.I hard-timeout ( 12 hours )
or if it is already logged out if it is past the
.I loggedout-keep-time. ( 2 hours )
If it is past these timeouts, the login cookie is deleted, as well as any
corresponding Kerberos tickets. When monster encounters a service
cookie, it looks up the corresponding login cookie and checks for
timeouts as specified above. If any are encountered, it deletes the
login cookie as well as the service cookie. If the service cookie's
corresponding login cookie has already been deleted, the service cookie is
also deleted.
.sp
If replication is enabled, monster also updates the other cosignds with
changes in the time stamps or state of its database of login cookies. It does
this through the use of the
.B TIME
command. See cosignd(8).
.SH STATS LOGGING
Upon each pass, Monster logs a line that contains the total number of
login and service cookies analyzed during the pass, and also notes how
many of these cookies were deleted. The line appears as follows:
.sp
STATS MONSTER: 1/6 login 2/4 service
.sp
This means Monster analyzed 6 login cookies and 4 service cookies, and
deleted 1 login cookie and 2 service cookies.
.SH TERMINOLOGY
.TP 19
.B login-cookie
in the form of "cosign=kl32kj42b...42b4kj2k", contains the user name,
ip address , realm, login time and optionally, the path to the user's
Kerberos V and/or Kerberos IV tickets. In addition, the login cookie
keeps state as to whether the user is logged in or logged out.
.TP 19
.B service-cookie
in the form of "cosign-campusmail=98ad9898fdg...798dfg9as", contains a
line that links to the user's login cookie.
.sp
.SH OPTIONS
.TP 19
.B \-d
debug mode. Does not disassociate from controlling tty.
.TP 19
.BI \-F\  syslog-facility
specifies which syslog-facilty to log messages to. Default is daemon.
.TP 19
.BI \-h\  replication-host
hostname to replicate to. This "turns on" monster's state propagation.
.TP 19
.BI \-H\  hard-timeout
specifies, in seconds, the absolute or hard timeout for any login cookie, by
default 12 hours.
.TP 19
.BI \-i\  idle-cache-timeout
specifies, in seconds, the timeout for removal of an idle login cookie, by
default 4.5 hours. This is the defualt idle time and grey window from cosignd,
and the default loggedout_cache time in monster.
.TP 19
.BI \-I\  timestamp-pushing-interval
specifies, in seconds, how long monster should sleep between passes, by
default 2 minutes.
.TP 19
.BI \-L\  syslog-level
specifies which syslog-level to log messages at. Default it info.
.TP 19
.BI \-l\  loggedout-cache
specifies, in seconds, how long to store logged out cookies in the
database before they are deleted, by default 2 hours.
.TP 19
.BI \-p\  port 
specifies the port of the cosign server, by default
.BR 6663 .
.TP 19
.B \-V
displays the version of 
.B  monster
and exits.
.TP 19
.BI \-x\  ca-dir
Certificate authority's public certificate directory, by default _COSIGN_TLS_CADIR.
.TP 19
.BI \-y\  cert-pem-file
Server's public certificate, by default _COSIGN_TLS_CERT.
.TP 19
.BI \-z\  private-key-file
Server's private key, by default _COSIGN_TLS_KEY.
.sp
.SH CONFIG FILE OPTIONS
The follow configuration file options are recognised. Configuration file options override the default compile time settings, but are in turn overridden by any command line options as specified. For the default values see the appropriate command line option which can override it.
.TP 19
.BI cosigncadir
The location where CA certificates in hashed form after expected to reside for validating client ceriticates. This is overridden by the
.B \-x
command line option
.TP 19
.BI cosigncert
The location of the SSL server certificate for use by this program. This is overridden by the
.B \-y
command line option
.TP 19
.BI cosignkey
The location of the SSL key file for use by this program. This is overridden by the
.B \-z
command line option
.TP 19
.BI cosigndb
The cosign daemon directory, where all tickets (cosign= and cosign-serv=) will be stored. This is overridden by the
.B \-D
command line option
.SH SEE ALSO
.sp
http://weblogin.org, cosignd(8)


daemon/monster.c
/*
 * Copyright (c) 2004 Regents of The University of Michigan.
 * All Rights Reserved.  See COPYRIGHT.
 */

#include <sys/types.h>
#include <sys/time.h>
#include <sys/stat.h>
#include <sys/param.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <string.h>
#include <unistd.h>
#include <syslog.h>
#include <stdlib.h>
#include <stdio.h>
#include <netdb.h>
#include <fcntl.h>
#include <dirent.h>

#include <openssl/err.h>
#include <openssl/ssl.h>
#include <snet.h>

#include "cparse.h"
#include "mkcookie.h"
#include "logname.h"
#include "rate.h"
#include "monster.h"
#include "config.h"

/* idle_cache = (grey+idle) from cosignd, plus loggedout_cache here */
int		idle_cache = (60 * 30) +  (60 * 60 * 2) + (60 * 60 * 2);
int		interval = 60 * 2;
int		hard_timeout = 60 * 60 * 12;
int		loggedout_cache = 60 * 60 * 2;
int             debug = 0;
int		login_gone;
int		hashlen = 0;
extern char	*cosign_version;

int		login_total, login_sent, service_total, service_gone;

static void (*logger)( char * ) = NULL;

static int eat_cookie( char *, struct timeval *, time_t *, int * );
static void do_dir( char *, struct connlist *, struct timeval * );

char    	*cosign_dir = _COSIGN_DIR;
char		*cryptofile = _COSIGN_TLS_KEY;
char		*certfile = _COSIGN_TLS_CERT;
char		*cadir = _COSIGN_TLS_CADIR;
struct timeval	cosign_net_timeout = { 60 * 4, 0 };
unsigned short	cosign_port;

    static void
monster_configure()
{
    char	 *val;

    if (( val = cosign_config_get( COSIGNDBKEY )) != NULL ) {
	syslog( LOG_INFO, "config: overriding default DB location(%s)"
		" to config value of '%s'", cosign_dir, val );
	cosign_dir = val;
    }

    if (( val = cosign_config_get( COSIGNCADIRKEY )) != NULL ) {
	syslog( LOG_INFO, "config: overriding default CA dir(%s)"
		" to config value of '%s'", cadir, val );
	cadir = val;
    }

    if (( val = cosign_config_get( COSIGNCERTKEY )) != NULL ) {
	syslog( LOG_INFO, "config: overriding default ssl cert location(%s)"
		" to config value of '%s'", certfile, val );
	certfile = val;
    }

    if (( val = cosign_config_get( COSIGNKEYKEY )) != NULL ) {
	syslog( LOG_INFO, "config: overriding default ssl key location(%s)"
		" to config value of '%s'", cryptofile, val );
	cryptofile = val;
    }

    if (( val = cosign_config_get( COSIGNTIMEOUTKEY )) != NULL ) {
	syslog( LOG_INFO, "config: overriding default net timeout of (%lu)"
		" to config value of '%s'", cosign_net_timeout.tv_sec, val );
	cosign_net_timeout.tv_sec = atoi( val );
	cosign_net_timeout.tv_usec = 0;
    }

    if (( val = cosign_config_get( COSIGNPORTKEY )) != NULL ) {
	syslog( LOG_INFO, "config: overriding default port of (%d)"
		" to config value of '%s'", cosign_port, val );
	cosign_port = htons( atoi( val ));
    } else {
	cosign_port = htons( 6663 );
    }

    if (( val = cosign_config_get( COSIGNDBHASHLENKEY )) != NULL ) {
	syslog( LOG_INFO, "config: overriding default dbhashlen of (%d)"
		" to config value of '%s'", hashlen, val );
	hashlen = atoi( val );
    }
}


    int
main( int ac, char **av )
{
    struct timeval	tv, now;
    struct hostent	*he;
    struct connlist	*head = NULL,*new = NULL, *temp, *yacur = NULL;
    struct connlist	**tail = NULL, **cur;
    char		hostname[ MAXHOSTNAMELEN ];
    char		hashdir[ 3 ];
    char		*sixtyfourchars = "abcdefghijklmnopqrstuvwxyz"
    					"ABCDEFGHIJKLMNOPQRSTUVWXYZ"
					"0123456789+-";
    char		*prog, *line;
    int			c, i, err = 0;
    char           	*cosign_host = NULL;
    char		*cosign_conf = _COSIGN_CONF;
    char		*p, *q;
    int                 facility = _COSIGN_LOG, level = LOG_INFO;
    SSL_CTX		*m_ctx = NULL;
    extern int          optind;
    extern char         *optarg;

    if (( prog = strrchr( av[ 0 ], '/' )) == NULL ) {
	prog = av[ 0 ];
    } else {
	prog++;
    }


#define MONSTER_OPTS "c:dF:h:H:i:I:l:L:p:Vx:y:z:"
    while (( c = getopt( ac, av, MONSTER_OPTS )) != EOF ) {
	switch ( c ) {
	case 'c':
	    cosign_conf = optarg;
	    break;

        default:
            break;
	}
    }

    if ( cosign_config( cosign_conf ) < 0 ) {
	fprintf( stderr, "monster: cosign_config failed.\n" );
	exit( 1 );
    }
    monster_configure();

    /* reset optind to parse all other args */
    optind = 1;

    while (( c = getopt( ac, av, MONSTER_OPTS )) != EOF ) {
	switch ( c ) {
	case 'c':
	    break;

	case 'd' :		/* debug */
	    debug++;
	    break;

	case 'F' :              /* syslog facility */
	    if (( facility = syslogfacility( optarg )) == -1 ) {
		fprintf( stderr, "%s: %s: unknown syslog facility\n",
			prog, optarg );
		exit( 1 );
	    }
	    break;

	case 'h' :		/* host running cosignd */
	    cosign_host = optarg;
	    break;

	case 'H' :		/* hard timeout for all cookies */
	    hard_timeout = atoi( optarg );
	    break;

	case 'i' :              /* idle timeout in seconds*/
	    idle_cache = atoi( optarg );
	    break;

	case 'I' :              /* timestamp pushing interval*/
	    interval = atoi( optarg );
	    break;

	case 'l' :              /* how long to keep logged out cookies*/
	    loggedout_cache = atoi( optarg );
	    break;

	case 'L' :              /* syslog level */
	    if (( level = sysloglevel( optarg )) == -1 ) {
		fprintf( stderr, "%s: %s: unknown syslog level\n",
			prog, optarg );
		exit( 1 );
	    }
	    break;

	case 'p' :              /* TCP port */
	     cosign_port = htons( atoi( optarg ));
	     break;

	case 'V' :              /* version */
	    printf( "%s\n", cosign_version );
	    exit( 0 );

	case 'x' :		/* ca dir */
	    cadir = optarg;
	    break;
	
	case 'y' :		/* cert */
	    certfile = optarg;
	    break;
	
	case 'z' :		/* private key file */
	    cryptofile = optarg;
	    break;
	
	case '?':
            err++;
            break;

        default:
            err++;
            break;
	}
    }

    if ( err || optind != ac ) {
	fprintf( stderr, "Usage: monster [ -c conf ] [ -dV ] " );
	fprintf( stderr, "[ -F syslog-facility ] [ -h cosignd-host ] ");
	fprintf( stderr, "[ -H hard-timeout  ] [ -i idlecachetimeinsecs ] " );
	fprintf( stderr, "[ -I update-interval ] [ -l loggedoutcachetime ]  " );
	fprintf( stderr, "[ -L syslog-level] [ -p port ] [ -x ca-dir ] " );
	fprintf( stderr, "[ -y cert-file] [ -z private-key-file ]\n" );
	exit( -1 );
    }

    if ( cosign_host != NULL ) {
	if ( gethostname( hostname, sizeof( hostname )) < 0 ) {
	    perror( "gethostname" );
	    exit( 1 );
	}

	if (( he = gethostbyname( cosign_host )) == NULL ) {
	    fprintf( stderr, "host unknown: %s\n", cosign_host );
	    exit( 1 );
	}
	tail = &head;
	for ( i = 0; he->h_addr_list[ i ] != NULL; i++ ) {
	    if (( new = ( struct connlist * )
		    malloc( sizeof( struct connlist ))) == NULL ) {
		perror( "connlist build" );
		exit( 1 );
	    }

	    memset( &new->cl_sin, 0, sizeof( struct sockaddr_in ));
	    new->cl_sin.sin_family = AF_INET;
	    new->cl_sin.sin_port = cosign_port;
	    memcpy( &new->cl_sin.sin_addr.s_addr,
		    he->h_addr_list[ i ], (unsigned int)he->h_length );
	    new->cl_sn = NULL;
	    new->cl_last_time = 0;
	    *tail = new;
	    tail = &new->cl_next;
	}
	*tail = NULL;

	SSL_load_error_strings();
	SSL_library_init();

	if ( cosign_ssl( cryptofile, certfile, cadir, &m_ctx ) != 0 ) {
	    fprintf( stderr, "monster: ssl setup error\n" );
	    exit( 1 );
	}
    }

    if ( chdir( cosign_dir ) < 0 ) {
	perror( cosign_dir );
	exit( 1 );
    }

    /* Disassociate from controlling tty. */

    if ( !debug ) {
	int		i, fd, dt;

	switch ( fork()) {
	case 0 :
	    if ( setsid() < 0 ) {
		perror( "setsid" );
		exit( 1 );
	    }
	    if (( fd = open( "/", O_RDONLY, 0 )) < 0 ) {
		perror( "open" );
		exit( 1 );
	    }
	    dt = getdtablesize();
	    for ( i = 0; i < dt; i++ ) {
		if ( i != fd ) {			
		    (void)close( i );
		}
	    }
	    (void)dup2( fd, 0 );
	    (void)dup2( fd, 1 );
	    (void)dup2( fd, 2 );

	    (void)close( fd );
	    break;

	case -1 :
	    perror( "fork" );
	    exit( 1 );
	default :
	    exit( 0 );
	}
    }

    /*
     * Start logging.
     */
    openlog( prog, LOG_NOWAIT|LOG_PID, facility );
    setlogmask( LOG_UPTO( level ));

    syslog( LOG_INFO, "restart %s", cosign_version );

	for (;;) {

    sleep( interval );
    login_total = service_total = login_gone = service_gone = login_sent = 0;

    if ( gettimeofday( &now, NULL ) != 0 ){
	syslog( LOG_ERR, "gettimeofday: %m" );
	exit( -1 );
    }

    /*
     * Usually, we'd write this as a nice neat for loop.  In this case,
     * since we have the ugly combination of a traversal and a possible
     * deletion, we use a while loop so we can better control the increment.
     */
    cur = &head;
    while ( *cur != NULL ) {
	if ( (*cur)->cl_sn == NULL ) {
	    if ( connect_sn( *cur, m_ctx, cosign_host, 1 ) != 0 ) {
		goto next;
	    }

	    snet_writef( (*cur)->cl_sn, "DAEMON %s\r\n", hostname );

	    tv = cosign_net_timeout;
	    if (( line = snet_getline_multi( (*cur)->cl_sn, logger, &tv ))
		    == NULL ) {
		syslog( LOG_ERR, "snet_getline_multi: 1: %m" );
		if (( snet_close( (*cur)->cl_sn )) != 0 ) {
		    syslog( LOG_ERR, "monster: snet_close: 2: %m" );
		}
		goto next;
	    }

	    if ( *line == '4' ) {
		if (( close_sn( *cur )) != 0 ) {
		    syslog( LOG_ERR, "close_sn: 3: %m" );
		}
		temp = *cur;
		*cur = (*cur)->cl_next;
		free( temp );
		/*
		 * we don't need to increment the loop in this case
		 * because the delete implicitly does.
		 */
		continue;

	    } else if ( *line != '2' ) {
		syslog( LOG_ERR, "getline: 4: %s", line );
		if (( close_sn( *cur )) != 0 ) {
		    syslog( LOG_ERR, "close_sn: 5: %m" );
		}
		goto next;
	    }
	}

	if ( snet_writef( (*cur)->cl_sn, "TIME\r\n" ) < 0 ) {
	    syslog( LOG_ERR, "snet_writef failed on TIME");
	    if ( snet_close( (*cur)->cl_sn ) != 0 ) {
		syslog( LOG_ERR, "snet_close: 6: %m" );
	    }
	    goto next;
	}

	tv = cosign_net_timeout;
	if (( line = snet_getline_multi( (*cur)->cl_sn, logger, &tv ))
		== NULL ) {
	    if ( !snet_eof( (*cur)->cl_sn )) {
		syslog( LOG_ERR, "snet_getline_multi: 7: %m" );
	    }
	    if ( snet_close( (*cur)->cl_sn ) != 0 ) {
		syslog( LOG_ERR, "snet_close: 8: %m" );
	    }
	    goto next;
	}

	if ( *line != '3' ) {
	    syslog( LOG_ERR, "snet_getline_multi: 9: %s", line );
	    if ( snet_close( (*cur)->cl_sn ) != 0 ) {
		syslog( LOG_ERR, "snet_close: 10: %m" );
	    }
next:
	    (*cur)->cl_sn = NULL;
	}
	cur = &(*cur)->cl_next;
    }

    switch ( hashlen ) {
    case 0 :
	do_dir( ".", head, &now );
	break;
    
    case 1 :
	for ( p = sixtyfourchars; *p != '\0'; p++ ) {
	    hashdir[ 0 ] = *p;
	    hashdir[ 1 ] = '\0';
	    do_dir( hashdir, head, &now );
	}
	break;

    case 2 :
	for ( p = sixtyfourchars; *p != '\0'; p++ ) {
	    for ( q = sixtyfourchars; *q != '\0'; q++ ) {
		hashdir[ 0 ] = *p;
		hashdir[ 1 ] = *q;
		hashdir[ 2 ] = '\0';
		do_dir( hashdir, head, &now );
	    }
	}

    default :
	syslog( LOG_ERR, "Illegal hashlen %d", hashlen );
	exit( 1 );
    }

    for ( yacur = head; yacur != NULL; yacur = yacur->cl_next ) {
	if ( yacur->cl_sn != NULL ) {
	    snet_writef( yacur->cl_sn, ".\r\n" );
	    if (( line = snet_getline_multi( yacur->cl_sn, logger, &tv ))
		     == NULL ) {
		if ( !snet_eof( yacur->cl_sn )) {
		    syslog( LOG_ERR, "snet_getline_multi: 13: %m" );
		}
		if ( snet_close( yacur->cl_sn ) != 0 ) {
		    syslog( LOG_ERR, "snet_close: 14: %m" );
		}
		yacur->cl_sn = NULL;
		continue;
	    }
	    if ( *line != '2' ) {
		syslog( LOG_ERR, "snet_getline_multi: 15: %m" );
		if ( snet_close( yacur->cl_sn ) != 0 ) {
		    syslog( LOG_ERR, "snet_close: 16: %m" );
		}
		yacur->cl_sn = NULL;
		continue;
	    }
	    yacur->cl_last_time = now.tv_sec;
	}

    }
    syslog( LOG_NOTICE, "STATS MONSTER: %d/%d/%d login %d/%d service",
	    login_gone, login_sent, login_total, service_gone, service_total );
	} /* end forever loop */
}

    void
do_dir( char *dir, struct connlist *head, struct timeval *now )
{
    DIR			*dirp;
    struct dirent	*de;
    char		path[ MAXPATHLEN ];
    char		lpath[ MAXPATHLEN ];
    struct connlist	*yacur;
    char                login[ MAXCOOKIELEN ];
    int			state = 0;
    time_t		itime = 0;
    int			rc;

    if (( dirp = opendir( dir )) == NULL ) {
	syslog( LOG_ERR, "%s: %m", cosign_dir);
	exit( 1 );
    }
    while (( de = readdir( dirp )) != NULL ) {
	/* is a login cookie */

	if ( mkcookiepath( NULL, hashlen, de->d_name,
		path, sizeof( path )) < 0 ) {
	    continue;
	}

	if ( strncmp( de->d_name, "cosign=", 7 ) == 0 ) {
	    login_total++;

	    if (( rc = eat_cookie( path, now, &itime, &state )) < 0 ) {
		syslog( LOG_ERR, "eat_cookie failure: %s", path );
		continue;
	    }
	    for ( yacur = head; yacur != NULL; yacur = yacur->cl_next ) {
		if (( itime > yacur->cl_last_time ) &&
			( yacur->cl_sn != NULL )) {
		    login_sent++;
		    if ( snet_writef( yacur->cl_sn, "%s %d %d\r\n",
			    de->d_name, itime, state ) < 0 ) {
			if ( snet_close( yacur->cl_sn ) != 0 ) {
			    syslog( LOG_ERR, "snet_close: 11: %m" );
			}
			yacur->cl_sn = NULL;
			continue;
		    }
		}
	    }
	} else if ( strncmp( de->d_name, "cosign-", 7 ) == 0 ) {
	    service_total++;
	    if ( service_to_login( path, login ) != 0 ) {
		continue;
	    }

	    if ( mkcookiepath( NULL, hashlen, login,
		    lpath, sizeof( lpath )) < 0 ) {
		syslog( LOG_ERR, "do_dir: mkcookiepath error." );
		exit( 1 );
	    }

	    if (( rc = eat_cookie( lpath, now, &itime, &state )) < 0 ) {
		syslog( LOG_ERR, "eat_cookie failure: %s", login );
		continue;
	    }
	    if ( rc == 0 ) {
		if ( unlink( path ) != 0 ) {
		    syslog( LOG_ERR, "%s: 12: %m", path );
		}
		service_gone++;
	    }
	} else {
	    continue;
	}
    }
    closedir( dirp );
}

    int
eat_cookie( char *name, struct timeval *now, time_t *itime, int *state )
{
    struct cinfo	ci = { 0, 0, "\0","\0","\0", "\0","\0", 0, };
    int			rc, create = 0;
    extern int		errno;


    /* -1 is a serious error
     * 0 means the cookie was deleted
     * 1 means still good and time was updated
     */

    if (( rc = read_cookie( name, &ci )) < 0 ) {
	syslog( LOG_ERR, "read_cookie error: %s", name );
	return( -1 );
    }

    /* login cookie gave us an ENOENT so we think it's gone */
    if ( rc == 1 ) {
	return( 0 );
    }

    /* logged out plus extra non-fail overtime */
    if ( !ci.ci_state && (( now->tv_sec - ci.ci_itime ) > loggedout_cache )) {
	goto delete_stuff;
    }

    /* idle out, plus gray window, plus non-failover */
    if (( now->tv_sec - ci.ci_itime )  > idle_cache ) {
	goto delete_stuff;
    }

    /* hard timeout */
    create = atoi( ci.ci_ctime );
    if (( now->tv_sec - create )  > hard_timeout ) {
	goto delete_stuff;
    }

    *itime = ci.ci_itime; 
    *state = ci.ci_state;
    return( 1 );

delete_stuff:

    /* remove krb5 ticket and login cookie */
    if ( strcmp( ci.ci_krbtkt, "\0" ) != 0 ) {
	if ( unlink( ci.ci_krbtkt ) != 0 ) {
	    syslog( LOG_ERR, "%s: %m", ci.ci_krbtkt );
	}
    }
    if ( unlink( name ) != 0 ) {
	syslog( LOG_ERR, "%s: %m", name );
    } 
    login_gone++;

    return( 0 );
}
